File: samples/csharp/Advanced/ImageInferWithOpenCV/OpenCVTools.cs
using System.Collections.Generic;
using System;
using OpenCvSharp;

namespace MMind.DL
{
    class OpenCVTools
    {
        /// <summary>
        /// Converts Mat to <see cref="MMindImage">MMindImage</see>.
        /// </summary>
        /// <param name="cvMat">The Mat data structure of OpenCV.</param>
        /// <returns>See <see cref="MMindImage">MMindImage</see> for details.</returns>
        public static MMindImage cvMatConvertToMMindImage(Mat cvMat)
        {
            MMindImage image = new MMindImage();
            image.Width = cvMat.Width;
            image.Height = cvMat.Height;
            image.Channel = cvMat.Channels();
            int nbytes = image.Width * image.Height * image.Channel;
            image.Data = new byte[nbytes];
            unsafe
            {
                fixed (byte* dataPtr = image.Data)
                {
                    Buffer.MemoryCopy(cvMat.DataPointer, dataPtr, nbytes, nbytes);
                }
            }
            return image;
        }

        /// <summary>
        /// Converts MMindImage to Mat.
        /// </summary>
        /// <param name="image">See <see cref="MMindImage">MMindImage</see> for details.</param>
        /// <returns>The Mat image.</returns>
        public static Mat MMindImageConvertToCvMat(MMindImage image)
        {
            MatType type = MatType.CV_8UC3;
            if (image.Channel == 1)
                type = MatType.CV_8UC1;
            unsafe
            {
                fixed (byte* dataPtr = image.Data)
                {
                    Mat cvMat = new Mat(image.Height, image.Width, type, (IntPtr)dataPtr);
                    return cvMat;
                }
            }
        }

        /// <summary>
        /// Converts contours from the structure of List<List<int>> to Point[][].
        /// </summary>
        /// <param name="contours">Contour points in the List<List<int>> structure.</param>
        /// <returns>Contours in the Point[][] structure.</returns>
        public static Point[][] ContoursToCvContours(List<List<int>> contours)
        {
            Point[][] pointsArray = new Point[contours.Count][];
            for (int i = 0; i < contours.Count; i++)
            {
                List<int> intPair = contours[i];
                pointsArray[i] = new Point[intPair.Count / 2];

                for (int j = 0; j < intPair.Count; j += 2)
                {
                    int x = intPair[j];
                    int y = intPair[j + 1];
                    pointsArray[i][j / 2] = new Point(x, y);
                }
            }
            return pointsArray;
        }
    }
}

File: samples/cpp/Advanced/ImageInferWithOpenCV/OpenCVTools.cpp
#include "OpenCVTools.h"

cv::Mat MMindImageToCvMat(const mmind::dl::MMindImage& image)
{
    cv::Mat imageTemp = {image.height, image.width, CV_MAKETYPE(image.depth, image.channel),
                         image.data.get()};
    return imageTemp.clone();
}

mmind::dl::MMindImage cvMatToMMindImage(const cv::Mat& mat)
{
    std::size_t matSize = mat.cols * mat.rows * mat.channels();
    unsigned char* ptr = new unsigned char[matSize];
    memcpy(ptr, mat.data, matSize);
    auto data = std::shared_ptr<unsigned char>(ptr, [](unsigned char* p) { delete[] p; });
    return {mat.cols, mat.rows, mat.channels(), mat.depth(), data};
}

cv::Point MMindPointToCvPoint(const mmind::dl::MMindPoint& point)
{
    return {static_cast<int>(point.x), static_cast<int>(point.y)};
}

std::vector<std::vector<cv::Point>> MMindContourToCvContour(const mmind::dl::MMindContour& contour)
{
    std::vector<std::vector<cv::Point>> cvContour;
    cvContour.reserve(contour.contourNum);
    for (const auto& points : contour.contourPoints) {
        std::vector<cv::Point> cvPoints;
        cvPoints.reserve(points.size());
        for (const auto& point : points) {
            cvPoints.emplace_back(MMindPointToCvPoint(point));
        }
        cvContour.push_back(cvPoints);
    }
    return cvContour;
}